flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/expressions/converter/CustomizedConvertRule.java

```diff
@@ -0,0 +1,361 @@
+		Boolean removeLeading = extractValue(removeLeadingExpr, Boolean.class);
+		ValueLiteralExpression removeTrailingExpr = (ValueLiteralExpression) children.get(1);
+		Boolean removeTrailing = extractValue(removeTrailingExpr, Boolean.class);
+		RexNode trimString = context.toRexNode(children.get(2));
+		RexNode str = context.toRexNode(children.get(3));
+		Enum trimMode;
+		if (removeLeading && removeTrailing) {
+			trimMode = SqlTrimFunction.Flag.BOTH;
+		} else if (removeLeading) {
+			trimMode = SqlTrimFunction.Flag.LEADING;
+		} else if (removeTrailing) {
+			trimMode = SqlTrimFunction.Flag.TRAILING;
+		} else {
+			throw new IllegalArgumentException("Unsupported trim mode.");
+		}
+		return context.getRelBuilder().call(
+			FlinkSqlOperatorTable.TRIM,
+			context.getRelBuilder().getRexBuilder().makeFlag(trimMode),
+			trimString,
+			str);
+	}
+
+	private static RexNode convertGet(List<Expression> children, ConvertContext context) {
+		RexNode child = context.toRexNode(children.get(0));
+		ValueLiteralExpression keyLiteral = (ValueLiteralExpression) children.get(1);
+		Optional<Integer> indexOptional = ExpressionUtils.extractValue(keyLiteral, String.class).map(
+			child.getType().getFieldNames()::indexOf);
+		// Note: never replace the following code with :
+		// int index = indexOptional.orElseGet(() -> extractValue(keyLiteral, Integer.class));
+		// Because the logical in `orElseGet` always executed no matter whether indexOptional is present or not.
+		int index;
+		index = indexOptional.orElseGet(() -> extractValue(keyLiteral, Integer.class));
+		return context.getRelBuilder().getRexBuilder().makeFieldAccess(child, index);
+	}
+
+	private static RexNode convertIn(List<Expression> children, ConvertContext context) {
+		Expression headExpr = children.get(1);
+		if (headExpr instanceof TableReferenceExpression) {
+			QueryOperation tableOperation = ((TableReferenceExpression) headExpr).getQueryOperation();
+			RexNode child = context.toRexNode(children.get(0));
+			return RexSubQuery.in(((FlinkRelBuilder) context.getRelBuilder())
+					.queryOperation(tableOperation).build(),
+				ImmutableList.of(child));
+		} else {
+			List<RexNode> child = context.toRexNodes(children);
+			return context.getRelBuilder().call(FlinkSqlOperatorTable.IN, child);
+		}
+	}
+
+	private static RexNode convertReinterpretCast(List<Expression> children, ConvertContext context) {
+		RexNode child = context.toRexNode(children.get(0));
+		TypeLiteralExpression type = (TypeLiteralExpression) children.get(1);
+		RexNode checkOverflow = context.toRexNode(children.get(2));
+		return context.getRelBuilder().getRexBuilder().makeReinterpretCast(
+			context.getTypeFactory().createFieldTypeFromLogicalType(
+				type.getOutputDataType().getLogicalType().copy(child.getType().isNullable())),
+			child,
+			checkOverflow);
+	}
+
+	private interface Conversion {
+		RexNode convert(List<Expression> children, ConvertContext context);
```

Comment:
How about changing that to:
`RexNode convert(CallExpression children, ConvertContext context);`?

This would solve the comments:
`// TODO get type from CallExpression directly`

In the current interface we simplify just a rather simple call: `call.getChildren`

Reply:
It should be, now we have removed the `UnresolvedCallExpression`, we can get all output types.

Reply:
But we still use PlannerExpressions to infer output type, that will lead to wrong precision of DecimalType.
So I will keep this TODO until we finish Expression type inference.